syntax = "proto3";
package Phantom.Scalable.Grpc.Internal;
import "google/protobuf/any.proto";
import "Phantom.Scalable/PhantomScalableGrpc.proto";

message PaxosBallotNumber {
    uint64 EpochNumber = 1;
    uint64 BallotNumber = 2;
}

message Phase1aMessage {
    PaxosBallotNumber BallotNumber = 1;
}

message Phase2aMessage {
}

message LearnMessage {
}

message Phase1bMessage {
}

message Phase2bMessage {
}

message NakMessage {
}

message InternalLockManagerOperation{
    oneof PayloadType {
        LockManagerOperation LockManagerOperation = 1;
    }
}

message InternalLockManagerResult {
    oneof PayloadType {
        LockManagerResult LockManagerResult = 1;
    }
}

message InternalOperation {
    oneof PayloadType {
        google.protobuf.Any Payload = 1;
        InternalLockManagerOperation LockManagerPayload = 3;
    }
}

message InternalOperationResult {
    TransactionOutcome TransactionOutcome = 1;
    oneof PayloadType {
        google.protobuf.Any Payload = 2;
        Fault Fault = 3;
        InternalLockManagerResult LockManagerPayload = 4;
    }
}

message InternalOperationInformation {
}

message InternalResourceManager {
    string InternalResourceManagerName = 1;
}

message InternalResource {
    string InternalResourceName = 1;
}

message Participant {
    InternalResourceManager InternalResourceManager = 1;
    InternalResource InternalResource = 2;
}

message InternalOperationIdentifier {
    string UniqueIdentifier = 1;
    Participant Participant = 2;
}

message ScalablePaxosRequestMessage {
    oneof MessageType {
        Phase1aMessage Phase1a = 1;
        Phase2aMessage Phase2a = 2;
        LearnMessage Learn = 3;
    }
}

message ScalablePaxosResponseMessage {
    oneof MessageType {
        Fault Fault = 1;
        Phase1bMessage Phase1b = 2;
        Phase2bMessage Phase2b = 3;
        NakMessage Nak = 4;
    }
}

service ScalablePeerToPeerService {
    rpc ProcessPaxosMessage(
        ScalablePaxosRequestMessage)
    returns (
        ScalablePaxosResponseMessage);
}

message DataComponent {
    oneof DataType {
        string String = 1;
        bytes Bytes = 2;
        google.protobuf.Any Any = 3;
        int64 Int64 = 4;
    }
}

message DataComponentString {
    repeated DataComponent Component = 1;
}

enum RelationshipType {
    Transitive = 0;
}

message InternalOperationInterferenceRelationship {
    RelationshipType RelationshipType = 1;
    DataComponentString Relationship = 2;
}

message LockContentKey {
    uint64 LockNumber = 1;
    uint64 Version = 2;
}

message LockContentValue {
    bytes Data = 1;
}

message LockMetadataKey {
    string LockName = 1;
}

message MutableLockMetadata {
    uint64 Version = 1;
}

message LockMetadataValue {
    uint64 LockNumber = 1;
    MutableLockMetadata Prepared = 2;
    MutableLockMetadata Committed = 3;
}
